network-config-analyzer/IstioSidecarYamlParser.py

```diff
@@ -0,0 +1,233 @@
+            # Services in the specified namespace matching dnsName will be exposed.
+            namespace, dns_name = self._validate_and_partition_host_format(host)
+            host_peers |= self._get_peers_from_host_namespace(namespace)
+            self._validate_dns_name_pattern(dns_name)
+            if host_peers:  # if there are services in the specified namespace
+                if '*' not in dns_name:  # * means all services in namespace, all already in host_peers
+                    host_peers &= self._get_peers_from_host_dns_name(dns_name)
+            res_peers |= host_peers
+
+        return IstioSidecarRule(res_peers)
+
+    def parse_policy(self):
+        """
+        Parses the input object to create a IstioSidecar object
+        :return: a IstioSidecar object with proper PeerSets
+        :rtype: IstioSidecar
+        """
+        if not isinstance(self.policy, dict):
+            self.syntax_error('type of Top ds is not a map')
+        if self.policy.get('kind') != 'Sidecar':
+            return None  # Not a Sidecar object
+        api_version = self.policy.get('apiVersion')
+        if 'istio' not in api_version:
+            return None  # apiVersion is not properly set
+        valid_keys = {'kind': [1, str], 'apiVersion': [1, str], 'metadata': [1, dict], 'spec': [0, dict]}
+        self.check_fields_validity(self.policy, 'Sidecar', valid_keys,
+                                   {'apiVersion': ['networking.istio.io/v1beta1']})
+        metadata = self.policy['metadata']
+        allowed_metadata_keys = {'name': [1, str], 'namespace': [0, str]}
+        self.check_fields_validity(metadata, 'metadata', allowed_metadata_keys)
+        self.namespace = self.peer_container.get_namespace(metadata.get('namespace', 'default'))
+        res_policy = IstioSidecar(metadata['name'], self.namespace)
+
+        if 'spec' not in self.policy or self.policy['spec'] is None:
+            self.warning('spec is missing or null in Sidecar ' + res_policy.full_name())
+            return res_policy
+
+        sidecar_spec = self.policy['spec']
+        # currently, supported fields in spec are workloadSelector, ingress and egress
+        allowed_spec_keys = {'workloadSelector': [0, dict], 'ingress': [0, list], 'egress': [0, list],
+                             'outboundTrafficPolicy': [2, str]}
+        self.check_fields_validity(sidecar_spec, 'Sidecar spec', allowed_spec_keys)
+        res_policy.affects_egress = sidecar_spec.get('egress') is not None
+        res_policy.affects_ingress = sidecar_spec.get('ingress') is not None
+
+        workload_selector = sidecar_spec.get('workloadSelector')
+        if workload_selector is None:
+            res_policy.selected_peers = self.peer_container.get_all_peers_group()
+        else:
+            res_policy.selected_peers = self._parse_workload_selector(workload_selector)
+        # if sidecar's namespace is the root namespace, then it applies to all cluster's namespaces
```

Comment:
Is there a difference in Istio's spec for workload selector of Sidecar vs AuthorizationPolicy? 
Can we have a common function for parsing this selector for both resources?

Reply:
* from the docs of istio, workload selector of both the authorization policy and sidecar are the same.
I committed a common function in a new class IstioGenericYamlParser .
however, while according to the docs , the "matchlabels" field is required for both objects (the selector may not be null or empty)
the "old" code of this method in IstioPolicyYamlParser, treated the field as optional. 
I will install and test the behavior of the fields on a live cluster to confirm what conditions should be included in the def

* i can move more parts from the "parse_policy" def  as "def"s to IstioGenericYamlParser , - like parsing the general fields (metadata, kind, apiversion ...)

Reply:
done